File: Problems/Algorithms/62. Unique Paths/UniquePaths.kt
class Solution {
    fun uniquePaths(m: Int, n: Int): Int {
        val dp: Array<IntArray> = Array(n) { IntArray(m) { 0 } }
        dp[0][0] = 1
        
        for (i in 0..n-1) {
            for (j in 0..m-1) {
                dp[i][j] += if (i > 0) dp[i-1][j] else 0
                dp[i][j] += if (j > 0) dp[i][j-1] else 0
            }
        }
        
        return dp[n-1][m-1]
    }
}

File: Problems/Algorithms/62. Unique Paths/unique_paths.rs
impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
        (m.max(n)..m+n-1)
            .zip(1..m.min(n))
            .fold(1, |acc, (num, den)| (acc * num as i128) / den as i128) as i32
    }
}

